
syntax = "proto3";

import "google/protobuf/empty.proto";

package mpc.leath;

service LeathRPC {

rpc pre_setup (google.protobuf.Empty) returns (preSetupMessage) {}

// Setup
rpc setup (SetupMessage) returns (SetupMessage) {}

// Share
rpc share (ShareRequestMessage) returns (google.protobuf.Empty) {}
rpc batch_share (stream ShareRequestMessage) returns (batchShareReply) {}

// Update
rpc reconstruct (ReconstructRequestMessage) returns (ReconstructReply) {}
rpc batch_reconstruct (stream ReconstructRequestMessage) returns (stream ReconstructReply) {}
rpc bulk_reconstruct (ReconstructRangeMessage) returns (stream ReconstructReply) {}

}

message preSetupMessage
{
    bytes G = 1;
    bytes H = 2;
    bytes N = 3; // modular for range proof
}

message SetupMessage
{
    uint64 msg_id = 1;
    bytes msg = 2;
}

message ShareRequestMessage
{
    uint64 value_id = 1;
    bytes value_share = 2;
    bytes  mac_share = 3;
}

message batchShareReply
{
    uint64 counter = 1;
}

message ReconstructRequestMessage
{
    uint64 value_id = 1;
}

message ReconstructRangeMessage
{
    uint64 begin_id = 1;
    uint64 end_id = 2;
}

message ReconstructReply
{
    uint64 value_id = 1;
    bytes value_share = 2;
    bytes  mac_share = 3;
}